Local names linking

xal-0 opened the pull request:
# Overview

This PR overhauls the way linking works in Julia, both in the JIT and AOT.  The point is to enable us to generate LLVM IR that depends only on the source IR, eliminating both nondeterminism and statefulness.  This serves two purposes.  First, if the IR is predictable, we can cache compile objects using the bitcode hash as a key, like how the ThinLTO cache works.  #58592 was an early experiment along these lines.  Second, we can reuse work that was done in a previous session, like pkgimages, but for the JIT.

We accomplish this by generating names that are unique only within the current LLVM module, removing most uses of the `globalUniqueGeneratedNames` counter. The replacement for `jl_codegen_params_t`, `jl_codegen_output_t`, represents a Julia "translation unit", and tracks the information we'll need to link the compiled module into the running session.  When linking, we manipulate the JITLink [LinkGraph](https://llvm.org/docs/JITLink.html#linkgraph) (after compilation) instead of renaming functions in the LLVM IR (before).

## Example

```
julia> @noinline foo(x) = x + 2.0
       baz(x) = foo(foo(x))

       code_llvm(baz, (Int64,); dump_module=true, optimize=false)
```

Nightly:
```llvm
[...]
@"+Core.Float64#774" = private unnamed_addr constant ptr @"+Core.Float64#774.jit"
@"+Core.Float64#774.jit" = private alias ptr, inttoptr (i64 4797624416 to ptr)

; Function Signature: baz(Int64)
;  @ REPL[1]:2 within `baz`
define double @julia_baz_772(i64 signext %"x::Int64") #0 {
top:
  %pgcstack = call ptr @julia.get_pgcstack()
  %0 = call double @j_foo_775(i64 signext %"x::Int64")
  %1 = call double @j_foo_776(double %0)
  ret double %1
}

; Function Attrs: noinline optnone
define nonnull ptr @jfptr_baz_773(ptr %"function::Core.Function", ptr noalias nocapture noundef readonly %"args::Any[]", i32 %"nargs::UInt32") #1 {
top:
  %pgcstack = call ptr @julia.get_pgcstack()
  %0 = getelementptr inbounds i8, ptr %"args::Any[]", i32 0
  %1 = load ptr, ptr %0, align 8
  %.unbox = load i64, ptr %1, align 8
  %2 = call double @julia_baz_772(i64 signext %.unbox)
  %"+Core.Float64#774" = load ptr, ptr @"+Core.Float64#774", align 8
  %Float64 = ptrtoint ptr %"+Core.Float64#774" to i64
  %3 = inttoptr i64 %Float64 to ptr
  %current_task = getelementptr inbounds i8, ptr %pgcstack, i32 -152
  %"box::Float64" = call noalias nonnull align 8 dereferenceable(8) ptr @julia.gc_alloc_obj(ptr %current_task, i64 8, ptr %3) #5
  store double %2, ptr %"box::Float64", align 8
  ret ptr %"box::Float64"
}
[...]
```

Diff after this PR.  Notice how each symbol gets the lowest possible integer suffix that will make it unique to the module, and how the two specializations for `foo` get different names:
```diff
@@ -4,18 +4,18 @@
 target triple = "arm64-apple-darwin24.6.0"
 
-@"+Core.Float64#774" = external global ptr
+@"+Core.Float64#_0" = external global ptr
 
 ; Function Signature: baz(Int64)
 ;  @ REPL[1]:2 within `baz`
-define double @julia_baz_772(i64 signext %"x::Int64") #0 {
+define double @julia_baz_0(i64 signext %"x::Int64") #0 {
 top:
   %pgcstack = call ptr @julia.get_pgcstack()
-  %0 = call double @j_foo_775(i64 signext %"x::Int64")
-  %1 = call double @j_foo_776(double %0)
+  %0 = call double @j_foo_0(i64 signext %"x::Int64")
+  %1 = call double @j_foo_1(double %0)
   ret double %1
 }
 
 ; Function Attrs: noinline optnone
-define nonnull ptr @jfptr_baz_773(ptr %"function::Core.Function", ptr noalias nocapture noundef readonly %"args::Any[]", i32 %"nargs::UInt32") #1 {
+define nonnull ptr @jfptr_baz_0(ptr %"function::Core.Function", ptr noalias nocapture noundef readonly %"args::Any[]", i32 %"nargs::UInt32") #1 {
 top:
   %pgcstack = call ptr @julia.get_pgcstack()
@@ -23,7 +23,7 @@
   %1 = load ptr, ptr %0, align 8
   %.unbox = load i64, ptr %1, align 8
-  %2 = call double @julia_baz_772(i64 signext %.unbox)
-  %"+Core.Float64#774" = load ptr, ptr @"+Core.Float64#774", align 8
-  %Float64 = ptrtoint ptr %"+Core.Float64#774" to i64
+  %2 = call double @julia_baz_0(i64 signext %.unbox)
+  %"+Core.Float64#_0" = load ptr, ptr @"+Core.Float64#_0", align 8
+  %Float64 = ptrtoint ptr %"+Core.Float64#_0" to i64
   %3 = inttoptr i64 %Float64 to ptr
   %current_task = getelementptr inbounds i8, ptr %pgcstack, i32 -152
@@ -39,8 +39,8 @@
 
 ; Function Signature: foo(Int64)
-declare double @j_foo_775(i64 signext) #3
+declare double @j_foo_0(i64 signext) #3
 
 ; Function Signature: foo(Float64)
-declare double @j_foo_776(double) #4
+declare double @j_foo_1(double) #4
 
 attributes #0 = { "frame-pointer"="all" "julia.fsig"="baz(Int64)" "probe-stack"="inline-asm" }
```

## List of changes
- Many sources of statefulness and nondeterminism in the emitted LLVM IR have been eliminated, namely:
  - Function symbols defined for CodeInstances
  - Global symbols referring to data on the Julia heap
  - Undefined function symbols referring to invoked external CodeInstances

- `jl_codeinst_params_t` has become `jl_codegen_output_t`.  It now represents one Julia "translation unit".  More than one CodeInstance can be emitted to the same `jl_codegen_output_t`, if desired, though in the JIT every CI gets its own right now.  One motivation behind this is to allow us to emit code on multiple threads and avoid the bitcode serialize/deserialize step we currently do, if that proves worthwhile.
  
  When we are done emitting to a `jl_codegen_output_t`, we call `.finish()`, which discards the intermediate state and returns only the LLVM module and the info needed for linking (`jl_linker_info_t`).

- The new `JLMaterializationUnit` wraps emitting Julia LLVM modules and the associated `jl_linker_info_t`.  It informs ORC that we can materialize symbols for the CIs defined by that output, and picks globally unique names for them. When it is materialized, it resolves all the call targets and generates trampolines for CodeInstances that are invoked but have the wrong calling convention, or are not yet compiled.

- We now postpone linking decisions to after codegen whenever possible.  For example, `emit_invoke` no longer tries to find a compiled version of the CodeInstance, and it no longer generates trampolines to adapt calling conventions.  `jl_analyze_workqueue`'s job has been absorbed into `JuliaOJIT::linkOutput`.

- Some `image_codegen` differences have been removed:
  - Codegen no longer cares if a compiled CodeInstance came from an image. During ahead-of-time linking, we generate thunk functions that load the address from the fvars table.

- In `jl_emit_native_impl`, emit every CodeInstance into one `jl_codegen_output_t`.  We now defer the creation of the `llvm::Linker` for llvmcalls, which has construction cost that grows with the size of the destination module, until the very end.

- RTDyld is removed completely, since we cannot control linking like we can with JITLink.  Since #60105, platforms that previous used the optimized memory manager now use the new one.

### General refactoring
- Adapt the `jl_callingconv_t` enum from `staticdata.c` into `jl_invoke_api_t` and use it in more places.  There is one enumerator for each special `jl_callptr_t` function that can go in a CodeInstance's `invoke` field, as well as one that indicates an invoke wrapper should be there.  There is a convenience function for reading an invoke pointer and getting the API type, and vice versa.
- Avoid using magic string values, and try to directly pass pointers to LLVM `Function *` or ORC string pool entries when possible.

## Future work
- `DLSymOptimizer` should be mostly removed, in favour of emitting raw ccalls and redirecting them to the appropriate target during linking.

- We should support ahead-of-time linking multiple `jl_codegen_output_t`s together, in order to parallelize LLVM IR emission when compiling a system image.
      
-  We still pass strings to `emit_call_specfun_other`, even though the prototype for the function is now created by `jl_codegen_output_t::get_call_target`.  We should hold on to the calling convention info so it doesn't have to be recomputed.

xal-0 committed:
Add jl_invoke_api_t enum and use it in staticdata.c

Renumber jl_invoke_api_t

xal-0 committed:
Set JL_CI_FLAGS_SPECPTR_SPECIALIZED only on specsig in jl_update_all_fptrs

xal-0 committed:
[JIT] Switch to post-compile linking

Use JITLink everywhere

Rename jlcall_type, add jl_funcs_invoke_ptr

Move JLLinkingLayer into JuliaOJIT

Use jl_invoke_api_t elsewhere

Rename JL_INVOKE_JFPTR -> JL_INVOKE_SPECSIG

Put all special symbol names in one place

Add helper for specsig -> tojlinvoke (fptr1) and use it

Fix invariants for code_outputs

Document JIT invariants better; remove invalid assertions

Replace workqueue, partially support OpaqueClosure

Add JIT tests

Stop using strings so much

Don't create an LLVM::Linker unless necessary

Generate trampolines in aot_link_output

GCChecker annotations, misc changes

Re-add emit_always_inline

Get JLDebuginfoPlugin and eh_frame working again

Re-add OpaqueClosure MethodInstance global root

Fix GCChecker annotations

Clean up TODOs

Read dump compile

Use multiple threads in the JIT

Add PLT/GOT for external fns

Name Julia PLT GOT entries

Do emit_llvmcall_modules at the end

Suppress clang-tidy, static analyzer warnings

Keep temporary_roots alive during emit_always_inline

Mark pkg PLT thunks noinline

Don't attempt to emit inline codeinsts when IR is too large or missing

Improve thunk generation on x86

Fix infinite loop in emit_always_inline if inlining not possible

Use local names for global targets

Fix jl_get_llvmf_defn_impl cfunction hacks

xal-0 committed:
Add CodegenParams.unique_names so they can be enabled in llvmpasses tests

xal-0 committed:
Fix timing_print_module_names not capturing JL_TIMING_DEFAULT_BLOCK

vchuravy commented on src/jitlayers.cpp:
```diff
@@ -1223,6 +869,73 @@ std::unique_ptr<jitlink::JITLinkMemoryManager> createJITLinkMemoryManager() JL_N
+    static JLMaterializationUnit Create(JuliaOJIT &JIT, ObjectLinkingLayer &OL,
+                                        std::unique_ptr<jl_linker_info_t> Info,
+                                        std::unique_ptr<MemoryBuffer> Obj) JL_NOTSAFEPOINT
+    {
```
Nice! I have been wanting this for a long time!

Would it make sense to have a C-API for creating these? So that LLVM.jl could create them?

vchuravy replied:
Yeah I have been wanting to try and ORC based setup for GPUCompiler

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

adienes commented:
> eliminating both nondeterminism and the effect of redefining methods in the same session

there are several open issues observing inference changes when methods are redefined; does this PR affect those?

xal-0 committed:
Use hand-written PLT thunks

Unfortunately the "portable" LLVM way of generating thunks doesn't generate the
code we want.  Instead, on platforms where it makes sense, we'll steal the LLD
PLT thunk code, but in disassembled form.  At some point this should be moved to
after linking, where it can be in assembled form again.  Amusingly it will be
more portable in assembled form, because the assembler syntax for relocations
differs between object formats.

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Use fallback PLT for i386

xal-0 committed:
Don't add Attribute::Naked unless using inline assembly PLT thunk

xal-0 committed:
Use TCK_MustTail in fallback PLT thunk

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Remove jl_link_global emulation from jl_get_llvmf_defn_impl

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Re-enable DLSymOptimizer for the time being

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Re-add JL_NOTSAFEPOINT_LEAVE/ENTER annotations on optimizeDLSyms

xal-0 committed:
Remove redundant separators between name and unique number; fix names.jl llvmpass test

gbaraldi commented on base/reflection.jl:
```diff
@@ -183,19 +183,28 @@ struct CodegenParams
+    across all code generated with this flag set.  Intended for llvmpasses
+    tests.
+    """
+    unique_names::Cint
```
Why is this needed? Can't we adjust the names?

xal-0 committed:
One extra fix for the names.jl llvmpass test

vtjnash commented on src/codegen.cpp:
```diff
@@ -2845,6 +2831,7 @@ std::unique_ptr<Module> jl_create_llvm_module(StringRef name, LLVMContext &conte
 {
     ++ModulesCreated;
     auto m = std::make_unique<Module>(name, context);
+    m->setUwtable(UWTableKind::None); // TODO remove, testing
```
cleanup TODO and commented out sections

vtjnash commented on src/jitlayers.cpp:
```diff
@@ -2315,54 +2007,225 @@ void JuliaOJIT::enablePerfJITEventListener()
+        // after it has been compiled.
+        Out = std::make_unique<jl_codegen_output_t>("tojlinvoke", getDataLayout(),
+                                                    getTargetTriple());
+        Function *F = emit_tojlinvoke(CI, StringRef(), *Out);
```
We should review to make sure this cannot acquire any mutexes or safepoints, since that would cause a lock inversion violation with the LinkerMutex

(e.g. annotate the containing function JL_NOTSAFEPOINT to indicate that the caller is permitted to hold a lock)

vtjnash commented on src/jitlayers.cpp:
```diff
@@ -2315,54 +2007,225 @@ void JuliaOJIT::enablePerfJITEventListener()
+    }
+    else {
+        // TODO: Check invariants to make sure this can't happen.
+        abort();
```
Happens in the (unlikely) case that it predicted we would want to compile with specsig, then later the runtime changes its mind (e.g. because it later ends up inferring `@nospecialize` from intersecting methods, or `--compile=min` is in use, etc.)

vtjnash commented on src/aotcompile.cpp:
```diff
@@ -792,6 +615,157 @@ void *jl_create_native_impl(LLVMOrcThreadSafeModuleRef llvmmod, int trim, int ex
+    else {
+        // Generic fallback that may be inefficient but won't mangle registers.
+        auto FPtr = B.CreateAlignedLoad(PtrTy, GV, out.DL.getPointerABIAlignment(0));
+        auto Call = B.CreateCall(FTy, FPtr, {});
```
you probably do need to pass along the actual arguments to keep llvm happy

xal-0 replied:
I believe it's special-cased in the verifier:
> If the musttail call appears in a function with the "thunk" attribute and the caller and callee both have varargs, then any unprototyped arguments in register or memory are forwarded to the callee. Similarly, the return value of the callee is returned to the caller’s caller, even if a void return type is in use.

vtjnash commented on src/aotcompile.cpp:
```diff
@@ -792,6 +615,157 @@ void *jl_create_native_impl(LLVMOrcThreadSafeModuleRef llvmmod, int trim, int ex
+        Code = plt_asm_riscv32;
+    }
+
+    auto FTy = FunctionType::get(Type::getVoidTy(Ctx), !Code);
```
```suggestion
    auto FTy = CallSite->getFunctionType();
```
no reason to need to make llvm unhappy with us when we have the correct value right here

vtjnash commented on src/aotcompile.cpp:
```diff
@@ -792,6 +615,157 @@ void *jl_create_native_impl(LLVMOrcThreadSafeModuleRef llvmmod, int trim, int ex
+    AttrBuilder Attrs{Ctx};
+    Attrs.addAttribute(Attribute::NoInline);
+    Attrs.addAttribute(Attribute::NoUnwind);
+    Attrs.addAttribute("frame-pointer", "none");
```
This'll might the Profiler sometimes segfault if it happens to hit in this function, which would not be great

xal-0 replied:
For some reason I thought unwind info wasn't generated for PLT thunks, but it looks like it is.  (Resolving assuming your comment is referring to `NoUnwind` and not `frame-pointer=none`.)

xal-0 replied:
Do you think we should force it to have `frame-pointer=all` when we do the generic fallback, or is ok to leave it unset? What I meant with my earlier comment was that I compiled some C programs, saw they generated unwind info for PLT thunks, and so removed the NoUnwind attribute so we match that.

vtjnash commented on src/jitlayers.h:
```diff
@@ -55,14 +55,6 @@
 #endif
 // The sanitizers don't play well with our memory manager
 
-#if defined(JL_FORCE_JITLINK) || defined(_CPU_AARCH64_) || defined(HAS_SANITIZER)
```
Maybe merge this separately, so we can get feedback if this causes issues for anyone separately from the PR (and because github has a hard time dealing with a diff this moderately large)

xal-0 committed:
Remove Attribute::NoUnwind on new modules, generate for PLT thunks

xal-0 committed:
Allow inlining of PLT thunks if no efficient version is available

xal-0 commented on src/aotcompile.cpp:
```diff
@@ -792,6 +615,157 @@ void *jl_create_native_impl(LLVMOrcThreadSafeModuleRef llvmmod, int trim, int ex
+        B.CreateUnreachable();
+    }
+    else {
+        // Generic fallback that may be inefficient but won't mangle registers.
```
It won't hurt to allow inlining---the "inefficiency" I was getting at is that some backends produce a ton of code to save and restore all the registers, even when it isn't necessary. I also updated the comment.

xal-0 committed:
Generate trampolines for args1 -> specsig case in linkCallTarget

xal-0 committed:
Annotate JuliaOJIT to reflect that we safepoint in materilization

xal-0 committed:
Suppress silly -Werror=maybe-uninitialized

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Suppress GCChecker error on ExecutionSession::reportError

xal-0 committed:
Revert "Annotate JuliaOJIT to reflect that we safepoint in materilization"

This reverts commit b41456d2a1575916080a8a825e94c2383b9ff294.

xal-0 committed:
Release engine_lock, exit gc_safe region after compiling CIs

xal-0 committed:
Add JLTrampolineMaterializationUnit, move compilation to materilize

xal-0 committed:
Don't leave JuliaTaskDispatcher::work_until until all tasks finish

xal-0 committed:
Set codeinst->time_compile once again

xal-0 committed:
Use only jl_code_info_t from codeinfos in emit_always_inline (aotcompile)

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Use global TaskQueue in JuliaTaskDispatcher, help other threads

xal-0 committed:
Fail with better error when symbol lookup/compile fails; don't abort

xal-0 committed:
Simplify JuliaTaskDispatcher to avoid deadlock

xal-0 committed:
Enter GC safe state when waiting in JuliaTaskDispatcher; fix GCChecker

xal-0 committed:
Fix error message capturing Exports by reference (outside of lifetime)

xal-0 committed:
Re-enable optimizeDLSyms in the JIT

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Ignore sections of materialize() in GCChecker

xal-0 committed:
Use a better symbol for tojlinvoke trampolines

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Attempt to detect deadlocks in JuliaTaskDispatcher

xal-0 committed:
Call set_value on promise even when lookup fails

xal-0 committed:
Print error in JuliaOJIT::publishCIs

xal-0 committed:
Hold LinkerMutex when it would be wrong for another thread to see the CISymbol

xal-0 committed:
Hold lock when checking future readiness

xal-0 committed:
Remove broken deadlock detection from JuliaTaskDispatcher

xal-0 committed:
Add LinkerMutex to devdocs

xal-0 commented:
@nanosoldier `runtests()`

vtjnash commented on src/cgutils.cpp:
```diff
@@ -4665,7 +4665,7 @@ static Value *emit_defer_signal(jl_codectx_t &ctx)
 }
 
-#ifndef JL_NDEBUG
+#if 0
```
Don't leave dead code behind (or commented out code)

xal-0 committed:
Disable JuliaTaskDispatcher::shutdown()

xal-0 committed:
Remove unused code for assertions

xal-0 committed:
Merge branch 'master' into local-names-linking

xal-0 committed:
Remove TODO

xal-0 committed:
When allocating a new CodeInstance, remove it from CISymbols

xal-0 committed:
Remove misleading jl_emit_oc_wrapper comments

topolarity reviewed (commented):
A few very small comments. I still have one more dive to do, but the changes look coherent (and well-factored) to me. @xal-0 are there any areas you'd like extra eyes on, for correctness or otherwise?

It can be challenging to find all of the "interesting" new code in a large change like this, especially when ~half of it is re-factoring.

> I don't love the solution of deleting entries from CISymbols when a new CodeInstance is allocated, but I also can't think of a better alternative.

This sounds reasonable to me. The only alternative would presumably be a finalizer-based solution that cleans up the CISymbols at the destruction of CI1? That has its own deadlock / threading concerns, but it does mean that we can actually iterate `CISymbols` to get a view of the code in the JIT, which is nice.

> It is now okay for the CodeInstance to be garbage collected, if that is possible (currently, only possible for top-level thunks).

This might be what you're alluding to w.r.t. top-level code, but it sounds like GC-ing other CodeInstances would be problematic for other reasons, such as that `jl_linker_info_t` becomes invalid to dereference the `ci`'s as it does now.

topolarity commented on src/jitlayers.cpp:
```diff
@@ -2336,54 +2103,206 @@ void JuliaOJIT::enablePerfJITEventListener()
+    // TODO: replace this with a GOT/PLT mechanism that avoids the jl_invoke
+    // after it has been compiled.
+    //
+    // We also generate a tojlinvoke to handle args1 -> specsig.
```
Great, no complaints here. Just planning for TypedCallable.

xal-0 committed:
Add jl_jit_unregister_ci, and use it in jl_eval_thunk instead

xal-0 commented:
> This sounds reasonable to me. The only alternative would presumably be a finalizer-based solution that cleans up the CISymbols at the destruction of CI1? That has its own deadlock / threading concerns, but it does mean that we can actually iterate `CISymbols` to get a view of the code in the JIT, which is nice.

@vtjnash suggested removing CodeInstances from CISymbols before we unroot them, since it only happens in `jl_eval_thunk` anyway. I've kept the old `jl_register_ci` in `jl_new_codeinst` around too, but it's just an assertion that should catch if any CodeInstances were JITted and garbage collected without unregistering.

xal-0 committed:
Remove JuliaOJIT::addModule (unused)

topolarity commented on src/jitlayers.cpp:
```diff
@@ -1223,6 +826,172 @@ class JLMemoryUsagePlugin : public ObjectLinkingLayer::Plugin {
+    {
+        auto &ES = R->getExecutionSession();
+
+        // TODO: Tell GCChecker that materialize can have safepoints.
```
Ah fair enough.

Yeah, let's leave that one for later.

xal-0 committed:
Call jl_jit_unregister_ci even when throwing an exception

xal-0 committed:
Mark jl_jit_unregister_mi as JL_NOTSAFEPOINT

topolarity reviewed (approved):
This is definitely the right direction for the JIT, and as a bonus the changes seem to be a net improvement in terms of clarity + simplicity (thanks for that @xal-0).

No major issues stand out to me. An approval from @vtjnash would be more meaningful than mine, since I do not have the "big picture" of our JIT pipeline. Nonetheless, I've done my best to do a "local" review of the JIT logic. This is already moderately well-tested by bootstrap / PkgEval, so I say we land this now and find any remaining bugs / gaps in the wild.

xal-0 commented on src/jitlayers.h:
```diff
@@ -699,15 +863,48 @@ inline orc::ThreadSafeModule jl_create_ts_module(StringRef name, orc::ThreadSafe
 void fixupTM(TargetMachine &TM) JL_NOTSAFEPOINT;
 
-void optimizeDLSyms(Module &M) JL_NOTSAFEPOINT_LEAVE JL_NOTSAFEPOINT_ENTER;
+void optimizeDLSyms(Module &M);
```
We now call `optimizeDLSyms` from `JLMaterializationUnit::materialize`, which is allowed to safepoint.

xal-0 commented on src/jitlayers.cpp:
```diff
@@ -1223,6 +826,172 @@ class JLMemoryUsagePlugin : public ObjectLinkingLayer::Plugin {
+        for (auto [CI, _] : Out.linker_info->ci_funcs) {
+            JL_GC_PROMISE_ROOTED(CI);
+            jl_do_dump_compile(CI, end_time - start_time);
+        }
```
Agreed. I'd like to improve Tracy output in general; since stackless inference it is no longer very useful for determining why something takes so long to compile.

xal-0 committed:
Update some out-of-date comments

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into local-names-linking

xal-0 committed:
Don't store references to call_targets in queue, since we may rehash

xal-0 committed:
Check return value of JIT.compileModule for failure in JLMaterializationUnit

xal-0 committed:
Check invoke_api in CISymbols before returning from linkCallTarget

xal-0 committed:
In jl_codegen_output_t, don't keep reference to owned TSM

Refer to the module through get_TSM() always, so that we don't move a reference
to the owned_TSM of the old jl_codegen_output_t in the move constructor.

xal-0 committed:
Add assertion to aot_link_output for documentation reasons

xal-0 committed:
Unregister toplevel CodeInstances before invoking, add tests

xal-0 commented:
Thanks to @vtjnash for the MWE that crashed on the old `jl_jit_unregister_ci`-after-invoke strategy. There's now a test in `test/jit.jl` that reliably triggers the assertion.  We now unregister the CodeInstance immediately before calling invoke, which is done by a special `jl_invoke_oneshot` helper, called from `jl_eval_thunk`.

xal-0 committed:
Fix broken BuiltinInvokeTag check in get_item_for_reloc

xal-0 committed:
Call jl_do_dump_compile in GC-unsafe region

xal-0 committed:
Remove some unused variables

xal-0 committed:
Remove commented-out jl_using_gdb_jitevents in JL_DEBUG_BUILD

xal-0 committed:
In emit_always_inline, erase old decl when emitting inline version

xal-0 committed:
Fix GCChecker eror in JLMaterializationUnit::materialize (again)

xal-0 committed:
Save and restore errno, last_error in jl_invoke_oneshot

xal-0 committed:
Merge branch 'master' into local-names-linking

xal-0 committed:
Sleep after test_gc_codeinst()
